wallet/src/de/schildbach/wallet/service/WalletTransactionMetadataProvider.kt

```diff
@@ -326,10 +381,71 @@ class WalletTransactionMetadataProvider @Inject constructor(
         }
     }
 
+    private fun updateIcon(txId: Sha256Hash, iconUrl: String) {
+        val request = Request.Builder().url(iconUrl).get().build()
+        Constants.HTTP_CLIENT.newCall(request).enqueue(object: Callback {
+            override fun onFailure(call: Call, e: IOException) {
+                log.error("Failed to fetch the icon for url: $iconUrl", e)
+            }
+
+            override fun onResponse(call: Call, response: Response) {
+                response.body?.let {
+                    syncScope.launch {
+                        try {
+                            val bitmap = decodeBitmap(it)
+                            val icon = resizeIcon(bitmap)
+                            val imageData = getBitmapData(icon)
+                            val imageHash = Sha256Hash.of(imageData)
+
+                            iconBitmapDao.addBitmap(IconBitmap(imageHash, imageData, icon.height, icon.width))
+                            transactionMetadataDao.updateIconId(txId, imageHash)
+                        } catch (ex: Exception) {
+                            log.error("Failed to resize and save the icon for url: $iconUrl", ex)
+                        }
+                    }
+                }
+            }
+        })
+    }
+
+    private fun decodeBitmap(responseBody: ResponseBody): Bitmap {
+        return BufferedInputStream(responseBody.byteStream()).use { inputStream ->
+            return@use BitmapFactory.decodeStream(inputStream)
+        }
+    }
+
+    private fun resizeIcon(bitmap: Bitmap): Bitmap {
+        var width = bitmap.width
+        var height = bitmap.height
+        val destSize = 150.0
+
+        if (width > destSize || height > destSize) {
+            if (width < height) {
+                val scale = destSize / height
+                height = destSize.toInt()
+                width = (width * scale).toInt()
+            } else if (width > height) {
+                val scale = destSize / width
+                width = destSize.toInt()
+                height = (height * scale).toInt()
+            }
+        }
+
+        return Bitmap.createScaledBitmap(bitmap, width, height, false)
+    }
+
+    private fun getBitmapData(bitmap: Bitmap): ByteArray {
+        val outputStream = ByteArrayOutputStream()
+        outputStream.use { bitmap.compress(Bitmap.CompressFormat.PNG, 100, it) }
+
+        return outputStream.toByteArray()
+    }
+
```

Comment:
In the first version of this I was caching the icon loaded by Coil and then saving it, however, it proved to be not reliable. Now `TransactionMetadataProvider` will be responsible for loading the icon in background.

I figure we can leave these methods here unless they are needed anywhere else in the future, at which point they can be moved into a separate class that handles image operations.